# 2. Coursera - Python Data Structures

Exercise 1: Write a while loop that starts at the last character in the
string and works its way backwards to the first character in the string,
printing each letter on a separate line, except backwards.

In [7]:
frase = 'Hola Mundo'
for letra in reversed(frase):
    print(letra)
    
#or

lista = [print(letra) for letra in reversed(frase)]

o
d
n
u
M
 
a
l
o
H
o
d
n
u
M
 
a
l
o
H


Exercise 2: Given that fruit is a string, what does fruit[:] mean?

In [9]:
fruit = 'banana'
fruit[:] # All characters

'banana'

Exercise 3: Encapsulate this code in a function named count, and generalize it so that it accepts the string and the letter as arguments.

In [14]:
def count(word):
    count = 0
    for letter in word:
        if letter == 'a':
            count = count + 1
    return print(count)

count('banana')

3


Exercise 4: There is a string method called count that is similar to the function in the previous exercise. Read the documentation of this method at:
https://docs.python.org/library/stdtypes.html#string-methods
Write an invocation that counts the number of times the letter a occurs in “banana”.

In [15]:
text = 'banana'
text.count('a')

3

6.5 Write code using find() and string slicing (see section 6.10) to extract the number at the end of the line below. Convert the extracted value to a floating point number and print it out.

In [1]:
text = "X-DSPAM-Confidence:    0.8475";
text = text.replace(" ", "")
p = text.find(':')
print(float(text[p+1:]))

0.8475


Exercise 6: Read the documentation of the string methods at https://docs.python.org/library/stdtypes.html#string-methods

You might want to experiment with some of them to make sure you understand how they work. strip and replace are particularly useful.

The documentation uses a syntax that might be confusing. For example, in find(sub[, start[, end]]), the brackets indicate optional arguments.
So sub is required, but start is optional, and if you include start, then end is optional.

In [16]:
text = 'banana'
text.replace('a','e')

'benene'

In [17]:
fhand = open('mbox.txt')
print(fhand)

<_io.TextIOWrapper name='mbox.txt' mode='r' encoding='UTF-8'>


In [18]:
fhand = open('mbox-short.txt')
count = 0
for line in fhand:
    count = count + 1
print('Line Count:', count)

Line Count: 1910


In [19]:
fhand = open('mbox.txt')
count = 0
for line in fhand:
    count = count + 1
print('Line Count:', count)

Line Count: 132045


In [20]:
fhand.close()

In [28]:
fhand = open('mbox-short.txt')
inp = fhand.read() # read the whole file
print(len(inp))

print(inp[:20])

print(type(fhand))
print(dir(fhand))

fhand.close()

94626
From stephen.marquar
<class '_io.TextIOWrapper'>
['_CHUNK_SIZE', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__lt__', '__ne__', '__new__', '__next__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '_checkClosed', '_checkReadable', '_checkSeekable', '_checkWritable', '_finalizing', 'buffer', 'close', 'closed', 'detach', 'encoding', 'errors', 'fileno', 'flush', 'isatty', 'line_buffering', 'mode', 'name', 'newlines', 'read', 'readable', 'readline', 'readlines', 'reconfigure', 'seek', 'seekable', 'tell', 'truncate', 'writable', 'write', 'write_through', 'writelines']


In [27]:
fhand.close()

In [32]:
fhand = open('mbox-short.txt')
count = 0
for line in fhand:
    if line.startswith('From:'):
        count += 1
print(count)
fhand.close()

27


In [33]:
fhand = open('mbox-short.txt')
count = 0
for line in fhand:
    line = line.rstrip()
    if line.startswith('From:'):
        count += 1
print(count)
fhand.close()

27


In [38]:
fhand = open('mbox-short.txt')
for line in fhand:
    line = line.rstrip()
    if line.find('@uct.ac.za') == -1: continue
    print(line)
fhand.close()

From stephen.marquard@uct.ac.za Sat Jan  5 09:14:16 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to stephen.marquard@uct.ac.za using -f
From: stephen.marquard@uct.ac.za
Author: stephen.marquard@uct.ac.za
From david.horwitz@uct.ac.za Fri Jan  4 07:02:32 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
r39753 | david.horwitz@uct.ac.za | 2008-01-04 13:05:51 +0200 (Fri, 04 Jan 2008) | 1 line
From david.horwitz@uct.ac.za Fri Jan  4 06:08:27 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
From david.horwitz@uct.ac.za Fri Jan  4 04:49:08 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
From david.horw

In [41]:
fname = input('Enter the file name: ')
fhand = open(fname)
count = 0
for line in fhand:
    if line.startswith('Subject:'):
        count += 1
print('There were', count, 'subject lines in', fname)
fhand.close()

Enter the file name: mbox.txt
There were 1797 subject lines in mbox.txt


In [1]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
    exit()
count = 0
for line in fhand:
    if line.startswith('Subject:'):
        count += 1
print('There were', count, 'subject lines in', fname)

Enter the file name: mbox.txt
There were 1797 subject lines in mbox.txt


In [3]:
fout = open('output.txt', 'w')
print(fout)

<_io.TextIOWrapper name='output.txt' mode='w' encoding='UTF-8'>


In [4]:
line1 = "This here's the wattle,\n"
fout.write(line1)

24

In [5]:
line2 = 'the emblem of our land.\n'
fout.write(line2)

24

In [6]:
fout.close()

In [7]:
s = '1 2\t 3\n 4'
print(s)

1 2	 3
 4


In [8]:
print(repr(s))

'1 2\t 3\n 4'


7.11 Exercises

Exercise 1: Write a program to read through a file and print the contents of the file (line by line) all in upper case. Executing the program will look as follows:

python shout.py
Enter a file name: mbox-short.txt
FROM STEPHEN.MARQUARD@UCT.AC.ZA SAT JAN 5 09:14:16 2008
RETURN-PATH: <POSTMASTER@COLLAB.SAKAIPROJECT.ORG>
RECEIVED: FROM MURDER (MAIL.UMICH.EDU [141.211.14.90])<br>
BY FRANKENSTEIN.MAIL.UMICH.EDU (CYRUS V2.3.8) WITH LMTPA;<br>
SAT, 05 JAN 2008 09:14:16 -0500

You can download the file from www.py4e.com/code3/mbox-short.txt

In [13]:
try:
    file = open("mbox-short.txt")
except:
    print("Cannot open mbox-short.txt")
else:
    for line in file:
        line = line.rstrip()
        print(line.upper())


FROM STEPHEN.MARQUARD@UCT.AC.ZA SAT JAN  5 09:14:16 2008
RETURN-PATH: <POSTMASTER@COLLAB.SAKAIPROJECT.ORG>
RECEIVED: FROM MURDER (MAIL.UMICH.EDU [141.211.14.90])
	 BY FRANKENSTEIN.MAIL.UMICH.EDU (CYRUS V2.3.8) WITH LMTPA;
	 SAT, 05 JAN 2008 09:14:16 -0500
X-SIEVE: CMU SIEVE 2.3
RECEIVED: FROM MURDER ([UNIX SOCKET])
	 BY MAIL.UMICH.EDU (CYRUS V2.2.12) WITH LMTPA;
	 SAT, 05 JAN 2008 09:14:16 -0500
RECEIVED: FROM HOLES.MR.ITD.UMICH.EDU (HOLES.MR.ITD.UMICH.EDU [141.211.14.79])
	BY FLAWLESS.MAIL.UMICH.EDU () WITH ESMTP ID M05EEFR1013674;
	SAT, 5 JAN 2008 09:14:15 -0500
RECEIVED: FROM PAPLOO.UHI.AC.UK (APP1.PROD.COLLAB.UHI.AC.UK [194.35.219.184])
	BY HOLES.MR.ITD.UMICH.EDU ID 477F90B0.2DB2F.12494 ;
	 5 JAN 2008 09:14:10 -0500
RECEIVED: FROM PAPLOO.UHI.AC.UK (LOCALHOST [127.0.0.1])
	BY PAPLOO.UHI.AC.UK (POSTFIX) WITH ESMTP ID 5F919BC2F2;
	SAT,  5 JAN 2008 14:10:05 +0000 (GMT)
MESSAGE-ID: <200801051412.M05ECIAH010327@NAKAMURA.UITS.IUPUI.EDU>
MIME-VERSION: 1.0
CONTENT-TRANSFER-ENCODING: 7BIT
RE

Exercise 2: Write a program to prompt for a file name, and then read through the file and look for lines of the form:

X-DSPAM-Confidence: 0.8475

When you encounter a line that starts with “X-DSPAM-Confidence:” pull apart the line to extract the floating-point number on the line. Count these lines and then compute the total of the spam confidence values from these lines. When you reach the end of the file, print out the average spam confidence.

Enter the file name: mbox.txt
Average spam confidence: 0.894128046745

Enter the file name: mbox-short.txt
Average spam confidence: 0.750718518519

Test your file on the mbox.txt and mbox-short.txt files.

In [28]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    count = 0
    sum_num = 0
    for line in fhand:
        if line.startswith('X-DSPAM-Confidence:'):
            count += 1
            line = line.rstrip()
            position = line.find(':')
            sum_num += float(line[position+2:])
    print('Average spam confidence:',sum_num/count)

Enter the file name: mbox.txt
Average spam confidence: 0.8941280467445736


Exercise 3: Sometimes when programmers get bored or want to have a bit of fun, they add a harmless Easter Egg to their program. Modify
the program that prompts the user for the file name so that it prints a funny message when the user types in the exact file name “na na boo
boo”. The program should behave normally for all other files which exist and don’t exist. Here is a sample execution of the program:

python egg.py<br>
Enter the file name: mbox.txt<br>
There were 1797 subject lines in mbox.txt

python egg.py<br>
Enter the file name: missing.tyxt<br>
File cannot be opened: missing.tyxt

python egg.py
Enter the file name: na na boo boo
NA NA BOO BOO TO YOU - You have been punk'd!

We are not encouraging you to put Easter Eggs in your programs; this is just an exercise.

In [32]:
name = input('Enter the file name: ')
try:
    fhand = open(name)
except:
    if 'na na boo boo NA NA BOO':
        print("BOO TO YOU - You have been punk'd!")
    else:
        print('File cannot be opened:',fhand)
else:
    c = 0
    for line in fhand:
        if line.startswith('Subject:'):
            c += 1
    print("There were",c,"subject lines in",name)

Enter the file name: mbox.txt
There were 1797 subject lines in mbox.txt


In [52]:
fhand = open('mbox-short.txt')
for line in fhand:
    line = line.rstrip()
    if not line.startswith('From '): continue
    words = line.split()
    print(words[2])

Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu


In [55]:
a = 'banana'
b = 'banana'
a is b

True

In [56]:
a = [1, 2, 3]
b = [1, 2, 3]
a is b

False

In [57]:
a = [1, 2, 3]
b = a
b is a

True

In [58]:
b[0] = 17
print(a)

[17, 2, 3]


In [62]:
a = 'banana'
b = a
a is b

True

Exercise 1: Write a function called chop that takes a list and modifies it, removing the first and last elements, and returns None. Then write a function called middle that takes a list and returns a new list that contains all but the first and last elements.

In [71]:
def chop(t):
    t = t[1:-1]
    return None

lista = [1,2,3,4,5,6,7]
print(chop(lista))

def middle(t):
    return t[1:-1]

print(middle(lista))

None
[2, 3, 4, 5, 6]


Exercise 2: Figure out which line of the above program is still not properly guarded. See if you can construct a text file which causes the program to fail and then modify the program so that the line is properly guarded and test it to make sure it handles your new text file.

In [72]:
fhand = open('mbox-short.txt')
count = 0
for line in fhand:
    words = line.split()
    if len(words) == 0: continue
    if words[0] != 'From': continue
    print(words[2])

Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu


Exercise 3: Rewrite the guardian code in the above example without two if statements. Instead, use a compound logical expression using the or logical operator with a single if statement.

In [73]:
fhand = open('mbox-short.txt')
count = 0
for line in fhand:
    words = line.split()
    if len(words) == 0 or words[0] != 'From': continue
    print(words[2])

Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu


Exercise 4: Download a copy of the file www.py4e.com/code3/romeo.txt. Write a program to open the file romeo.txt and read it line by line. For each line, split the line into a list of words using the split function. For each word, check to see if the word is already in a list. If the word is not in the list, add it to the list. When the program completes, sort and print the resulting words in alphabetical order.

Enter file: romeo.txt<br>
['Arise', 'But', 'It', 'Juliet', 'Who', 'already',
'and', 'breaks', 'east', 'envious', 'fair', 'grief',
'is', 'kill', 'light', 'moon', 'pale', 'sick', 'soft',
'sun', 'the', 'through', 'what', 'window',
'with', 'yonder']

In [77]:
try:
    file = open('romeo.txt')
except:
    print('Error to open a',file)
else:
    t = []
    for line in file:
        words = line.split()
        for word in words:
            if word in t: continue
            t.append(word)
print(sorted(t))
            

['Arise', 'But', 'It', 'Juliet', 'Who', 'already', 'and', 'breaks', 'east', 'envious', 'fair', 'grief', 'is', 'kill', 'light', 'moon', 'pale', 'sick', 'soft', 'sun', 'the', 'through', 'what', 'window', 'with', 'yonder']


Exercise 5: Write a program to read through the mail box data and when you find line that starts with “From”, you will split the line into words using the split function. We are interested in who sent the message, which is the second word on the From line.

From stephen.marquard@uct.ac.za Sat Jan 5 09:14:16 2008

You will parse the From line and print out the second word for each From line, then you will also count the number of From (not From:) lines and print out a count at the end. This is a good sample output with a few lines removed:

python fromcount.py<br>
Enter a file name: mbox-short.txt<br>
stephen.marquard@uct.ac.za<br>
louis@media.berkeley.edu<br>
zqian@umich.edu

[...some output removed...]

ray@media.berkeley.edu<br>
cwen@iupui.edu<br>
cwen@iupui.edu<br>
cwen@iupui.edu<br>
There were 27 lines in the file with From as the first word

In [82]:
file = input('Enter a file name: ')
try:
    file = open(file)
except:
    print('Error to open a',file)
else:
    c = 0
    for line in file:
        if line.startswith('From:') == True: continue
        if line.startswith('From'):
            words = line.split()
            print(words[1])
            c += 1
    print('There were',c,'lines in the file with From as the first word')

Enter a file name: mbox-short.txt
stephen.marquard@uct.ac.za
louis@media.berkeley.edu
zqian@umich.edu
rjlowe@iupui.edu
zqian@umich.edu
rjlowe@iupui.edu
cwen@iupui.edu
cwen@iupui.edu
gsilver@umich.edu
gsilver@umich.edu
zqian@umich.edu
gsilver@umich.edu
wagnermr@iupui.edu
zqian@umich.edu
antranig@caret.cam.ac.uk
gopal.ramasammycook@gmail.com
david.horwitz@uct.ac.za
david.horwitz@uct.ac.za
david.horwitz@uct.ac.za
david.horwitz@uct.ac.za
stephen.marquard@uct.ac.za
louis@media.berkeley.edu
louis@media.berkeley.edu
ray@media.berkeley.edu
cwen@iupui.edu
cwen@iupui.edu
cwen@iupui.edu
There were 27 lines in the file with From as the first word


Exercise 6: Rewrite the program that prompts the user for a list of numbers and prints out the maximum and minimum of the numbers at the end when the user enters “done”. Write the program to store the numbers the user enters in a list and use the max() and min() functions to compute the maximum and minimum numbers after the loop completes.

Enter a number: 6<br>
Enter a number: 2<br>
Enter a number: 9<br>
Enter a number: 3<br>
Enter a number: 5<br>
Enter a number: done<br>
Maximum: 9.0<br>
Minimum: 2.0

In [85]:
t = []
while True:
    num = input('Enter a number: ')
    if num == 'done': break
    else: t.append(float(num))
print('Maximum:',max(t))
print('Minimum:',min(t))

Enter a number: 5
Enter a number: done
Maximum: 5.0
Minimum: 5.0


Exercise 1: Download a copy of the file www.py4e.com/code3/words.txt

Write a program that reads the words in words.txt and stores them as keys in a dictionary. It doesn’t matter what the values are. Then you can use the in operator as a fast way to check whether a string is in the dictionary.

In [104]:
try:
    file = open("words.txt")
except:
    print("Error to open", file)
else:
    words = []
    for line in file:
        line = line.rstrip()
        for word in line.split():
            words.append(word)
    #words = set(words)
    words_dic = dict.fromkeys(words, 0)
    print(words_dic)

{'Writing': 0, 'programs': 0, 'or': 0, 'programming': 0, 'is': 0, 'a': 0, 'very': 0, 'creative': 0, 'and': 0, 'rewarding': 0, 'activity': 0, 'You': 0, 'can': 0, 'write': 0, 'for': 0, 'many': 0, 'reasons': 0, 'ranging': 0, 'from': 0, 'making': 0, 'your': 0, 'living': 0, 'to': 0, 'solving': 0, 'difficult': 0, 'data': 0, 'analysis': 0, 'problem': 0, 'having': 0, 'fun': 0, 'helping': 0, 'someone': 0, 'else': 0, 'solve': 0, 'This': 0, 'book': 0, 'assumes': 0, 'that': 0, '{\\em': 0, 'everyone}': 0, 'needs': 0, 'know': 0, 'how': 0, 'program': 0, 'once': 0, 'you': 0, 'program,': 0, 'will': 0, 'figure': 0, 'out': 0, 'what': 0, 'want': 0, 'do': 0, 'with': 0, 'newfound': 0, 'skills': 0, 'We': 0, 'are': 0, 'surrounded': 0, 'in': 0, 'our': 0, 'daily': 0, 'lives': 0, 'computers': 0, 'laptops': 0, 'cell': 0, 'phones': 0, 'think': 0, 'of': 0, 'these': 0, 'as': 0, 'personal': 0, 'assistants': 0, 'who': 0, 'take': 0, 'care': 0, 'things': 0, 'on': 0, 'behalf': 0, 'The': 0, 'hardware': 0, 'current-day': 0

In [107]:
word = 'brontosaurus'
d = {}
for c in word:
    if c not in d:
        d[c] = 1
    else:
        d[c] += 1
print(d)

{'b': 1, 'r': 2, 'o': 2, 'n': 1, 't': 1, 's': 2, 'a': 1, 'u': 2}


In [128]:
word = 'brontosaurus'
d = {}
for c in word:
    d[c] = d.get(c,0)+1
print(d)

{'b': 1, 'r': 2, 'o': 2, 'n': 1, 't': 1, 's': 2, 'a': 1, 'u': 2}


In [127]:
try:
    file = open('words.txt','r').read().split('\n')
except:
    print("Error to open", file)
else:
    text = "".join(file)
    d = {}
    for c in text:
        if c.isalpha():
            if c not in d:
                d[c] = 1
            else:
                d[c] += 1
    print(d)
    aux = {k: v for k, v in sorted(d.items(), key=lambda item: item[1])}
    print(aux)

{'W': 4, 'r': 59, 'i': 55, 't': 80, 'n': 72, 'g': 28, 'p': 23, 'o': 90, 'a': 70, 'm': 30, 's': 58, 'v': 13, 'e': 89, 'y': 20, 'c': 21, 'd': 28, 'w': 25, 'Y': 1, 'u': 45, 'f': 21, 'k': 16, 'l': 40, 'b': 10, 'h': 35, 'T': 2, 'q': 1, 'x': 3, 'O': 1, 'I': 2}
{'Y': 1, 'q': 1, 'O': 1, 'T': 2, 'I': 2, 'x': 3, 'W': 4, 'b': 10, 'v': 13, 'k': 16, 'y': 20, 'c': 21, 'f': 21, 'p': 23, 'w': 25, 'g': 28, 'd': 28, 'm': 30, 'h': 35, 'l': 40, 'u': 45, 'i': 55, 's': 58, 'r': 59, 'a': 70, 'n': 72, 't': 80, 'e': 89, 'o': 90}


In [133]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    counts = dict()
    for line in fhand:
        words = line.split()
        for word in words:
            counts[word] = counts.get(word,0)+1
    print(counts)

Enter the file name: romeo.txt
{'But': 1, 'soft': 1, 'what': 1, 'light': 1, 'through': 1, 'yonder': 1, 'window': 1, 'breaks': 1, 'It': 1, 'is': 3, 'the': 3, 'east': 1, 'and': 3, 'Juliet': 1, 'sun': 2, 'Arise': 1, 'fair': 1, 'kill': 1, 'envious': 1, 'moon': 1, 'Who': 1, 'already': 1, 'sick': 1, 'pale': 1, 'with': 1, 'grief': 1}


In [134]:
counts = { 'chuck' : 1 , 'annie' : 42, 'jan': 100}
for key in counts:
    print(key, counts[key])

chuck 1
annie 42
jan 100


In [135]:
counts = { 'chuck' : 1 , 'annie' : 42, 'jan': 100}
t = list(counts.keys())
print(t)
t.sort()
for key in t:
    print(key, counts[key])

['chuck', 'annie', 'jan']
annie 42
chuck 1
jan 100


In [136]:
import string
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [139]:
import string

fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    counts = {}
    for line in fhand:
        line = line.rstrip()
        line = line.translate(line.maketrans('', '', string.punctuation))
        line = line.lower()
        words = line.split()
        for word in words:
            counts[word] = counts.get(word,0)+1
    print(counts)

Enter the file name: romeo-full.txt
{'romeo': 40, 'and': 42, 'juliet': 32, 'act': 1, '2': 2, 'scene': 2, 'ii': 1, 'capulets': 1, 'orchard': 2, 'enter': 1, 'he': 5, 'jests': 1, 'at': 9, 'scars': 1, 'that': 30, 'never': 2, 'felt': 1, 'a': 24, 'wound': 1, 'appears': 1, 'above': 6, 'window': 2, 'but': 18, 'soft': 1, 'what': 11, 'light': 5, 'through': 2, 'yonder': 2, 'breaks': 1, 'it': 22, 'is': 21, 'the': 34, 'east': 1, 'sun': 2, 'arise': 1, 'fair': 4, 'kill': 2, 'envious': 2, 'moon': 4, 'who': 5, 'already': 1, 'sick': 2, 'pale': 1, 'with': 8, 'grief': 2, 'thou': 32, 'her': 14, 'maid': 2, 'art': 7, 'far': 2, 'more': 9, 'than': 6, 'she': 9, 'be': 14, 'not': 18, 'since': 1, 'vestal': 1, 'livery': 1, 'green': 1, 'none': 1, 'fools': 1, 'do': 7, 'wear': 1, 'cast': 1, 'off': 1, 'my': 29, 'lady': 2, 'o': 11, 'love': 24, 'knew': 1, 'were': 9, 'speaks': 3, 'yet': 9, 'says': 1, 'nothing': 1, 'of': 20, 'eye': 2, 'discourses': 1, 'i': 61, 'will': 8, 'answer': 1, 'am': 7, 'too': 8, 'bold': 1, 'tis': 4,

<b>Exercise 2:</b> Write a program that categorizes each mail message by which day of the week the commit was done. To do this look for lines that start with “From”, then look for the third word and keep a running count of each of the days of the week. At the end of the program print out the contents of your dictionary (order does not matter).

<b>Sample Line:</b>

From stephen.marquard@uct.ac.za Sat Jan 5 09:14:16 2008

<b>Sample Execution:</b>

python dow.py<br>
Enter a file name: mbox-short.txt<br>
{'Fri': 20, 'Thu': 6, 'Sat': 1}

In [155]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    dias = {}
    for line in fhand:
        line.rstrip()
        if not line.startswith('From:') and line.startswith('From'):
            line = line.split()
            dias[line[2]] = dias.get(line[2],0)+1 
    print(dias)

Enter the file name: mbox-short.txt
{'Sat': 1, 'Fri': 20, 'Thu': 6}


<b>Exercise 3:</b> Write a program to read through a mail log, build a histogram using a dictionary to count how many messages have come from each email address, and print the dictionary.

Enter file name: mbox-short.txt<br>
{'gopal.ramasammycook@gmail.com': 1, 'louis@media.berkeley.edu': 3,<br>
'cwen@iupui.edu': 5, 'antranig@caret.cam.ac.uk': 1,<br>
'rjlowe@iupui.edu': 2, 'gsilver@umich.edu': 3,<br>
'david.horwitz@uct.ac.za': 4, 'wagnermr@iupui.edu': 1,<br>
'zqian@umich.edu': 4, <br>'stephen.marquard@uct.ac.za': 2,<br>
'ray@media.berkeley.edu': 1}

In [158]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    dias = {}
    for line in fhand:
        line.rstrip()
        if not line.startswith('From:') and line.startswith('From'):
            line = line.split()
            dias[line[1]] = dias.get(line[1],0)+1 
    print(dias)

Enter the file name: mbox-short.txt
{'stephen.marquard@uct.ac.za': 2, 'louis@media.berkeley.edu': 3, 'zqian@umich.edu': 4, 'rjlowe@iupui.edu': 2, 'cwen@iupui.edu': 5, 'gsilver@umich.edu': 3, 'wagnermr@iupui.edu': 1, 'antranig@caret.cam.ac.uk': 1, 'gopal.ramasammycook@gmail.com': 1, 'david.horwitz@uct.ac.za': 4, 'ray@media.berkeley.edu': 1}


<b>Exercise 4:</b> Add code to the above program to figure out who has the most messages in the file. After all the data has been read and the dictionary has been created, look through the dictionary using a maximum loop (see Chapter 5: Maximum and minimum loops) to find who has the most messages and print how many messages the person has.

Enter a file name: mbox-short.txt<br>
cwen@iupui.edu 5

Enter a file name: mbox.txt<br>
zqian@umich.edu 195

In [166]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    emails = {}
    for line in fhand:
        line.rstrip()
        if not line.startswith('From:') and line.startswith('From'):
            line = line.split()
            emails[line[1]] = emails.get(line[1],0)+1 
    aux = max(emails.values())
    for email in emails:
        if emails[email] == aux:
            print(email, aux)

Enter the file name: mbox.txt
zqian@umich.edu 195


<b>Exercise 5:</b> This program records the domain name (instead of the address) where the message was sent from instead of who the mail came from (i.e., the whole email address). At the end of the program, print out the contents of your dictionary.

python schoolcount.py<br>
Enter a file name: mbox-short.txt<br>
{'media.berkeley.edu': 4, 'uct.ac.za': 6, 'umich.edu': 7,<br>
'gmail.com': 1, 'caret.cam.ac.uk': 1, 'iupui.edu': 8}

In [12]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    domains = {}
    for line in fhand:
        line.rstrip()
        if not line.startswith('From:') and line.startswith('From'):
            line = line.split()
            domain = line[1].split('@')
            domains[domain[1]] = domains.get(domain[1],0)+1 
    print(domains)

Enter the file name: 
File cannot be opened: 


<img src=juego.jpeg align=left><BR>

In [13]:
for a in range(1,10):
    for b in range(0,10):
        for r in range(0,10):
            if ((a*100)+(b*10)+r)*3 == (r*100)+(r*10)+r:
                print('azul: ',a,'\nblanco: ',b,'\nrojo: ', r)

azul:  1 
blanco:  8 
rojo:  5


In [22]:
for i in range(1,334):
    u = i%10
    if i*3 == ((u*100)+(u*10)+u):
        d = int(((i%100)-u)/10)
        c = int((i -i%100)/100)
        print('azul: ',c,'\nblanco: ',d,'\nrojo: ', u)

azul:  1 
blanco:  8 
rojo:  5


In [5]:
import string
fhand = open('romeo-full.txt')
counts = dict()
for line in fhand:
    line = line.translate(str.maketrans('', '', string.punctuation))
    line = line.lower()
    words = line.split()
    for word in words:
        if word not in counts:
            counts[word] = 1
        else:
            counts[word] += 1
    
# Sort the dictionary by value
lst = list()
for key, val in list(counts.items()):
    lst.append((val, key))

lst.sort(reverse=True)

for key, val in lst[:10]:
    print(key, val)

61 i
42 and
40 romeo
34 to
34 the
32 thou
32 juliet
30 that
29 my
24 thee


<b>Exercise 1:</b> Revise a previous program as follows: Read and parse the “From” lines and pull out the addresses from the line. Count the number of messages from each person using a dictionary.

After all the data has been read, print the person with the most commits by creating a list of (count, email) tuples from the dictionary. Then sort the list in reverse order and print out the person who has the most commits.

Sample Line:<br>
From stephen.marquard@uct.ac.za Sat Jan
5 09:14:16 2008

Enter a file name: mbox-short.txt<br>
cwen@iupui.edu 5

Enter a file name: mbox.txt<br>
zqian@umich.edu 195

In [15]:
fname = input('Enter a file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    domains = dict()
    for line in fhand:
        line.rstrip()
        if line.startswith('From'):
            line = line.split()
            domain = line[1].split('@')
            domains[domain[1]] = domains.get(domain[1],0)+1 

    #Sort the dictionary by value
    lst=list()
    for key, val in list(domains.items()):
        lst.append((val, key))
    
    lst.sort(reverse=True)
    
    print(lst[0][1], lst[0][0])

Enter a file name: mbox-short.txt
iupui.edu 16


<b>Exercise 2:</b> This program counts the distribution of the hour of the day for each of the messages. You can pull the hour from the “From” line by finding the time string and then splitting that string into parts using the colon character. Once you have accumulated the counts for each hour, print out the counts, one per line, sorted by hour as shown below.

python timeofday.py
Enter a file name: mbox-short.txt<br>
04 3<br>
06 1<br>
07 1<br>
09 2<br>
10 3<br>
11 6<br>
14 1<br>
15 2<br>
16 4<br>
17 2<br>
18 1<br>
19 1<br>

In [31]:
fname = input('Enter a file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    hours = dict()
    for line in fhand:
        line.rstrip()
        if not line.startswith('From:') and line.startswith('From'):
            line = line.split()
            hour = line[5][:2]
            hours[hour] = hours.get(hour,0)+1 

    #Sort the dictionary by value
    lst=list()
    for key, val in list(hours.items()):
        lst.append((val, key))
    
    lst.sort(reverse=True)
    r = [print(key, val) for val, key in lst]

Enter a file name: mbox.txt
10 198
16 181
09 166
11 154
15 152
14 150
13 114
12 108
17 97
08 72
18 58
19 46
06 45
07 42
21 36
20 36
22 28
04 23
00 23
03 19
23 18
02 11
05 10
01 10


In [51]:
# Versión 2

fname = input('Enter a file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    hours = dict()
    for line in fhand:
        line.rstrip()
        if not line.startswith('From:') and line.startswith('From'):
            line = line.split()
            hour = line[5][:2]
            hours[hour] = hours.get(hour,0)+1 

    #Sort the dictionary by value
    lst=list()
    for key, val in list(hours.items()):
        lst.append((key, val))
    
    lst.sort()
    r = [print(val, key) for val, key in lst]

Enter a file name: mbox-short.txt
04 3
06 1
07 1
09 2
10 3
11 6
14 1
15 2
16 4
17 2
18 1
19 1


<b>Exercise 3:</b> Write a program that reads a file and prints the letters in decreasing order of frequency. Your program should convert all the input to lower case and only count the letters a-z. Your program should not count spaces, digits, punctuation, or anything other than the letters a-z. Find text samples from several different languages and see how letter frequency varies between languages. Compare your results with the tables at https://wikipedia.org/wiki/Letter_frequencies.

In [49]:
fname = input('Enter a file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
else:
    letters = dict()
    count = 0
    for line in fhand:
        line.rstrip()
        line = [x.lower() for x in line]
        line = ''.join(x for x in line if x.isalpha())
        for letter in line:
            letters[letter] = letters.get(letter,0)+1
            count += 1

    #Sort the dictionary by value
    lst=list()
    for key, val in list(letters.items()):
        lst.append((val, key))
    
    lst.sort(reverse=True)
    r = [print(key, '%.2f' %(val/count*100)) for val, key in lst]

Enter a file name: mbox.txt
e 9.60
a 8.44
i 7.52
o 7.34
t 7.19
r 6.70
s 6.56
c 5.66
u 5.05
n 4.53
m 4.23
p 4.00
d 3.52
l 3.44
h 2.47
k 2.07
b 2.03
v 1.78
f 1.76
g 1.63
j 1.22
y 1.14
w 1.11
x 0.81
q 0.10
z 0.09
